### PR TITLE
🔥 Feature (v3): Add buffered streaming support

### LINKED ISSUE #3127: 📝 [Proposal]: Add buffered streaming support
### Feature Proposal Description

This feature proposal proposes to add buffered streaming support to Fiber v3. By adding this feature, Fiber users can send shorter segments of content over persistent HTTP connections. This functionality is important for several web-based applications such as:

- Server-Side Events (SSE)
- Large File Downloads

To implement this feature, this would involve using the underlying [`*fasthttp.RequestCtx`](https://github.com/valyala/fasthttp/blob/master/server.go#L588) struct. Specifically, this feature can leverage the [`RequestCtx.SetBodyStreamWriter`](https://github.com/valyala/fasthttp/blob/master/server.go#L1506) method.

```go
// File: https://github.com/valyala/fasthttp/server.go
// Note: type StreamWriter func(w *bufio.Writer)
func (ctx *RequestCtx) SetBodyStreamWriter(sw StreamWriter) {
	ctx.Response.SetBodyStreamWriter(sw)
}
```

The `StreamWriter` function passed into `RequestCtx.SetBodyStreamWriter` describes what a handler should write to an endpoint's response body as a buffered stream.

Within Fiber's existing codebase, implementing this feature is the equivalent of:
```go
c.SendStream(fasthttp.NewStreamReader(func(w *bufio.Writer) {
    // ...
}))
```

It is also possible to implement this feature by directly using the fasthttp API:
```go
// From a gofiber/recipe example:
// https://github.com/gofiber/recipes/blob/master/sse/main.go#L59
c.Context().SetBodyStreamWriter(fasthttp.StreamWriter(func(w *bufio.Writer) {
    // ...
})
```

For the feature's API, please view **Feature Examples** below.

### Alignment with Express API

Using the Express API, buffered streaming is supported for `req` argument in an express route handler by using the `write`, and `end` methods. Below is an example usage of this feature for **Server-Side Events (SSE)**:

```javascript
app.get('/sse', function(req, res) {
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Connection', 'keep-alive');
    res.flushHeaders();
	
    // Send messages every 2 seconds
    let intervalIndex = 0;
    const intervalId = setInterval(function() {
        intervalIndex++;

        // Format message using Server-Side Event specification
        // `data: ${content}\n\n`
        const currTime = new Date().toLocaleTimeString();
        const msg = `data: Message: ${i} - the time is ${currTime}\n\n`;

        // Write the message to the response stream (auto-flushes)
        res.write(msg);

        // Handle errors during writing
        res.on('error', function(err) {
            console.error('Error writing SSE message:', err);
            clearInterval(intervalId);
            res.end(); // Close the connection
        });
    }, 2000);
	
    // Handle client disconnection
    req.on('close', function() {
        clearInterval(intervalId);
        res.end();
    });
});
```

By providing a way to support buffered streams in Fiber, this will help make transitioning from Express much easier for more advanced use cases such as SSE as shown above.

To view the same Express example in Fiber with this new feature, please view **Feature Examples**.

### HTTP RFC Standards Compliance

This feature does not modify how Fiber or fasthttp complies with HTTP RFC standards, as it just provides simplified access to an existing fasthttp feature.

### API Stability

This feature's API will be introduced in v3 as a separate `Ctx` method, augmenting its interface without modifying existing methods. This will ensure that there are no breaking changes to old and new features for Fiber v3. (For more information, please view **Feature Examples** below.)

Besides potential naming changes, this API feature will require very few changes in the future due to `fasthttp` methods it relies on have been stable since 9 years ago. Unless `fasthttp` makes big API changes to their [`fasthttp.StreamWriter`](https://github.com/valyala/fasthttp/blob/master/stream.go#L19) type, maintenance will be limited.

On the rare circumstance that the StreamWriter type does change, we can use an alias for that type and have no code change required. The only change that we would need is to update documentation.

```go
type StreamHandler fasthttp.StreamWriter
```

### Feature Examples

Function Signature:

```go
func (c Ctx) SendStreamWriter(streamWriter func(*bufio.Writer)) error
```

> **Note:** The name of this method is still being finalized. A few other options are:
> - WriteStream
> - SendWritableStream
> - SendBufferedStream

Use in Server-Side Events (SSE):
```go
// Modified version of SSE example in gofiber/recipes
// https://github.com/gofiber/recipes/blob/master/sse/main.go#L59
app.Get("/sse", func (c Ctx) error {
    c.Set("Content-Type", "text/event-stream")
    c.Set("Cache-Control", "no-cache")
    c.Set("Connection", "keep-alive")
    c.Set("Transfer-Encoding", "chunked")

    return c.SendStreamWriter(func (w *bufio.Writer) {
        intervalIndex := 0
        for {
            intervalIndex++

            // Format message using Server-Side Event specification
            // `data: ${content}\n\n`
            currTime := time.Now().Local()
            msg := fmt.Sprintf("%d - the time is %v", i, currTime)    
            fmt.Fprintf(w, "data: Message: %s\n\n", msg)    // write the message

            // Handle errors while writing (after flushing)
            err := w.Flush()    // flush the message
            if err != nil {
                break    // close the connection
            }
            time.Sleep(2 * time.Second)    // Send messages every 2 seconds
        }
    })
})
```

### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have searched for existing issues that describe my proposal before opening this one.
- [X] I understand that a proposal that does not meet these guidelines may be closed without explanation.

### PR DESCRIPTION
**Pull request by grivera64**
# Description

This feature adds buffered streaming support to Fiber v3 through the new Ctx method `SendStreamWriter`:

```go
func (c Ctx ) SendStreamWriter(streamWriter func(w *bufio.Writer))
```

`c.SendStreamWriter()` essentially is a wrapper for calling fasthttp's [`SetBodyStreamWriter`](https://github.com/valyala/fasthttp/blob/master/server.go#L1506) method. This feature wraps this method in the same way that [`c.SendStream()`](https://github.com/gofiber/fiber/blob/main/ctx.go#L1662) wraps fasthttp's [`SetBodyStream()`](https://github.com/valyala/fasthttp/blob/master/server.go#L1491).

With this feature, Fiber users can send shorter segments of content over persistent HTTP connections. This functionality is important for several web-based applications such as:

- Server-Side Events (SSE)
- Large File Downloads

For example, a simple self-contained SSE example using this new feature can be setup as the following:

```go
homeHtml := `<!DOCTYPE html>
<html>
    <body>
        <h1>SSE Messages</h1>
        <div id="result"></div>

        <script>
            if (typeof(EventSource) !== "undefined") {
                var source = new EventSource("http://127.0.0.1:3000/sse");
                source.addEventListener("my-event", function(event) {
                    document.getElementById("result").innerHTML += event.data + "<br>";
                });
            } else {
                document.getElementById("result").innerHTML = "Sorry, your browser does not support server-sent events...";
            }
        </script>
    </body>
</html>
`
app := fiber.New()
app.Get("/", func(c fiber.Ctx) error {
    c.Set("Content-Type", fiber.MIMETextHTMLCharsetUTF8)
    return c.SendString(homeHtml)
})

app.Get("/sse", func (c fiber.Ctx) error {
    c.Set("Content-Type", "text/event-stream")
    c.Set("Cache-Control", "no-cache")
    c.Set("Connection", "keep-alive")
    c.Set("Transfer-Encoding", "chunked")

    return c.Status(fiber.StatusOK).SendStreamWriter(func (w *bufio.Writer) {
        for {
            fmt.Fprintf(w, "event: my-event\ndata: %s\n\n", time.Now().Local())
            if err := w.Flush(); err != nil {
                return
            }
            time.Sleep(time.Second)
        }
    })
})
```

Fixes #3127 

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] Documentation update (changes to documentation)

### CURRENT STATUS

- [X] Features
  - [X] Add `Ctx.SendStreamWriter()` to ctx.go
- [X] Unit Tests
  - [X] Add `Test_Ctx_SendStream_Writer` to ctx_test.go
  - [X] Add `Test_Ctx_SendStreamWriter_Interrupted` to ctx_test.go

    > ~~⚠️ WARNING:~~ FIXED ✅
    > ~~This test gives a **race condition** warning when running with `-race`, but is stable and consistent~~
    > ~~via mutex/channels.~~

- [X] Documentation
  - [X] Add `Ctx.SendStreamWriter()` docs to docs/api/ctx.md
- [ ] Benchmarks

## Checklist

Before you submit your pull request, please make sure you meet these requirements:

- [X] Followed the inspiration of the Express.js framework for new functionalities, making them similar in usage.
- [X] Conducted a self-review of the code and provided comments for complex or critical parts.
- [X] Updated the documentation in the `/docs/` directory for [Fiber's documentation](https://docs.gofiber.io/).
- [X] Added or updated unit tests to validate the effectiveness of the changes or new features.
- [X] Ensured that new and existing unit tests pass locally with the changes.
- [X] Verified that any new dependencies are essential and have been agreed upon by the maintainers/community.
- [X] Aimed for optimal performance with minimal allocations in the new code.
- [ ] Provided benchmarks for the new code to analyze and improve upon.

### CODE REVIEW

**Comment by gaby** [comment]
@grivera64 Try this:

```go
func Test_Ctx_SendStreamWriter_Interrupted(t *testing.T) {
    t.Parallel()
    app := New()
    c := app.AcquireCtx(&fasthttp.RequestCtx{})

    var mutex sync.Mutex
    var wg sync.WaitGroup // WaitGroup to synchronize goroutines
    startChan := make(chan bool)

    interruptStreamWriter := func() {
        defer wg.Done() // Notify WaitGroup when done
        <-startChan
        time.Sleep(5 * time.Millisecond)
        mutex.Lock()
        c.Response().CloseBodyStream() //nolint:errcheck // It is fine to ignore the error
        mutex.Unlock()
    }

    wg.Add(1) // Increment WaitGroup counter before starting goroutine
    err := c.SendStreamWriter(func(w *bufio.Writer) {
        go interruptStreamWriter()

        startChan <- true
        for lineNum := 1; lineNum <= 5; lineNum++ {
            mutex.Lock()
            fmt.Fprintf(w, "Line %d\n", lineNum) //nolint:errcheck, revive // It is fine to ignore the error
            mutex.Unlock()

            if err := w.Flush(); err != nil {
                if lineNum < 3 {
                    t.Errorf("unexpected error: %s", err)
                }
                return
            }

            time.Sleep(1500 * time.Microsecond)
        }
    })
    require.NoError(t, err)

    wg.Wait() // Wait for the interruptStreamWriter to finish

    // Protect access to the response body with the mutex
    mutex.Lock()
    defer mutex.Unlock()
    require.Equal(t, "Line 1\nLine 2\nLine 3\n", string(c.Response().Body()))
}
```

**Comment by grivera64** [comment]
@gaby Thanks for the recommendation! Adding the Wait Group does remove the race error, but now I am getting an empty response body. I think this may be due to one of the following:

- Using `c.Context().CloseBodyStream()` may be freeing the response body internally, making `c.Context().Body()` no longer valid after waiting long enough. This seems strange, but is the most probable, as it is a race condition.
  - The fix for this would be to find a different method to initiate a client disconnect.
- The Wait Group isn't waiting long enough to allow `Flush()` to work. This is least probable, since I also tried increasing the counter to ensure that both `interruptStreamWriter` and `SendStreamWriter` finished executing, and the body returned is still empty:

```go
// go test -run Test_Ctx_SendStreamWriter_Interrupted
func Test_Ctx_SendStreamWriter_Interrupted(t *testing.T) {
	t.Parallel()
	app := New()
	c := app.AcquireCtx(&fasthttp.RequestCtx{})

	var mutex sync.Mutex
	var wg sync.WaitGroup
	startChan := make(chan bool)

	interruptStreamWriter := func() {
		wg.Add(1)
		defer wg.Done()
		<-startChan
		time.Sleep(5 * time.Millisecond)
		mutex.Lock()
		c.Response().CloseBodyStream() //nolint:errcheck // It is fine to ignore the error
		mutex.Unlock()
	}

	wg.Add(1)
	err := c.SendStreamWriter(func(w *bufio.Writer) {
		go interruptStreamWriter()
		defer wg.Done()

		startChan <- true
		for lineNum := 1; lineNum <= 5; lineNum++ {
			mutex.Lock()
			fmt.Fprintf(w, "Line %d\n", lineNum) //nolint:errcheck, revive // It is fine to ignore the error
			mutex.Unlock()

			if err := w.Flush(); err != nil {
				if lineNum < 3 {
					t.Errorf("unexpected error: %s", err)
				}
				return
			}

			time.Sleep(1500 * time.Microsecond)
		}
	})
	require.NoError(t, err)

	// Wait for StreamWriter and the goroutine to finish
	wg.Wait()
	mutex.Lock()
	require.Equal(t, "Line 1\nLine 2\nLine 3\n", string(c.Response().Body()))
	mutex.Unlock()
}
```

I will file an issue on [valyala/fasthttp](https://github.com/valyala/fasthttp) to ask how to mock client disconnections for this test. If it's not possible, we could remove this test case, as most of the other tests do not test for client disconnection issues.

**Comment by grivera64** [comment]
Hey all, after reading a bit more on Fiber and Fasthttp documentation, I believe that the race condition is due to `fasthttp.(*Response)` not being concurrent-safe.

I tried using a modified version of app's `Test()` method, that closes an underlying `net.Conn` connection (I modified `testConn` too to make `Close()` block further writes but allow reads). This seemed to do the same as `c.Response().CloseBodyStream()`, but still gave the race condition issue.

This is the new race condition warning:

```txt
WARNING: DATA RACE
Read at 0x00c00016c130 by goroutine 9:
  github.com/gofiber/fiber/v3.(*testConn).Write()
      /fiber/helpers.go:628 +0x6b
  bufio.(*Writer).Flush()
      /usr/local/go/src/bufio/bufio.go:639 +0xee
  github.com/valyala/fasthttp.writeChunk()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/http.go:2250 +0x10b
  github.com/valyala/fasthttp.writeBodyChunked()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/http.go:2170 +0xce
  github.com/valyala/fasthttp.(*Response).writeBodyStream()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/http.go:2066 +0x338
  github.com/valyala/fasthttp.(*Response).Write()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/http.go:1967 +0x2c4
  github.com/valyala/fasthttp.writeResponse()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/server.go:2589 +0xb8
  github.com/valyala/fasthttp.(*Server).serveConn()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/server.go:2432 +0x1ead
  github.com/valyala/fasthttp.(*Server).ServeConn()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/server.go:2042 +0x154
  github.com/gofiber/fiber/v3.(*App).TestWithInterrupt.func1()
      /fiber/app.go:975 +0xde
...
```

This was the old warning when directly using `c.SendStreamWriter()` without a server:

```txt
github.com/valyala/fasthttp.(*Response).SetBodyStream()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/http.go:249 +0x4f
  github.com/valyala/fasthttp.(*Response).SetBodyStreamWriter()
      /go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/http.go:292 +0x64
...
```
I believe fixing this race warning in the cleanest way possible would require an upstream PR to fasthttp (most likely adding a mutex for `fasthttp.(*Response)` related  methods).

Based on Fiber's current codebase, there doesn't seem to be other interrupt tests (while other tests ignore output when the response times out). If this isn't something we should be testing for, we could just remove the interrupt test and keep the remaining tests.

What are your thoughts on this? Please let me know if you want to see the modified `app.Test()` code.

**Comment by grivera64** [comment]
After editing `app.Test()` a bit, I was able to get a working test without the race warning. I had to create an app method called `TestWithInterrupt()`:

```go
// TestWithInterrupt is used for internal debugging by passing a *http.Request with an interruptAfter duration.
func (app *App) TestWithInterrupt(req *http.Request, interruptAfter time.Duration) (*http.Response, error)
```

With this change, I was able to use that in the test case as written below as a fix:

```go
func Test_Ctx_SendStreamWriter_Interrupted_New(t *testing.T) {
	t.Parallel()
	app := New(Config{StreamRequestBody: true})
	app.Get("/", func(c Ctx) error {
		return c.SendStreamWriter(func(w *bufio.Writer) {
			for lineNum := 1; lineNum <= 5; lineNum++ {
				time.Sleep(time.Duration(lineNum) * time.Millisecond)

				fmt.Fprintf(w, "Line %d\n", lineNum) //nolint:errcheck, revive // It is fine to ignore the error
				if err := w.Flush(); err != nil {
					if lineNum <= 3 {
						t.Errorf("unexpected error: %s", err)
					}
					return
				}
			}
		})
	})
	resp, err := app.TestWithInterrupt(httptest.NewRequest(MethodGet, "/", nil), 8*time.Millisecond)
	require.NoError(t, err, "app.TestWithInterrupt(req)")
	body, err := io.ReadAll(resp.Body)
	require.NotNil(t, err)
	require.Equal(t, "Line 1\nLine 2\nLine 3\n", string(body))
}
```

Would you all like me to write an issue/PR for adding `app.TestWithInterrupt()` then use the function here after the merge?

**Comment by grivera64** [comment]
Hello all, I hope you all have been doing well. I created an issue for the TestWithInterrupt() I proposed above. Please check it out here:

#3149

**Comment by efectn** [comment]
@grivera64  can you update testcases using new app.Test config properties?

**Comment by grivera64** [comment]
> @grivera64 can you update testcases using new app.Test config properties?

For sure @efectn, I am working on using the new `app.Test()` for the unit test `Test_Ctx_SendStreamWriter_Interrupted`.

Should I also update the previous `Test_Ctx_SendStreamWriter` unit test to be consistent with the interrupted version, even though it still works as-is?

**Comment by efectn** [comment]
> > @grivera64 can you update testcases using new app.Test config properties?
> 
> For sure @efectn, I am working on using the new `app.Test()` for the unit test `Test_Ctx_SendStreamWriter_Interrupted`.
> 
> Should I also update the previous `Test_Ctx_SendStreamWriter` unit test to be consistent with the interrupted version, even though it still works as-is?

I think it's good as-is since app.Test is not used for that test

**Comment by ReneWerner87** [comment]
@grivera64 thanks for this feature
can you check our ai hints 
this could help to get the best out of it

File: docs/api/ctx.md [hunk 1]
```diff
@@ -1852,6 +1852,47 @@ app.Get("/", func(c fiber.Ctx) error {
 })
 ```
 
+## SendStreamWriter
```
Comment (ReneWerner87) [inline docs/api/ctx.md:1855]: please mention this new feature with example in our what's new section
(please not as a bullet point -> I still have to formulate it - in the right proportions like the other examples)

**Review by ReneWerner87 — APPROVED** [review]

File: ctx.go [hunk 2]
```diff
@@ -1671,6 +1672,13 @@ func (c *DefaultCtx) SendStream(stream io.Reader, size ...int) error {
 	return nil
 }
 
+// SendStreamWriter sets response body stream writer
+func (c *DefaultCtx) SendStreamWriter(streamWriter func(*bufio.Writer)) error {
```
[the remark this answers is not in the document]
Comment (efectn) [reply ctx.go:1676]: I think we should add an example/recipe about this feature after the PR get merged

File: docs/api/ctx.md (hunk 1 above)
Comment (grivera64) [reply docs/api/ctx.md:1855]: For sure, I have added an entry to /docs/whats_new.md for `SendStreamWriter`

**Comment by grivera64** [comment]
> @grivera64 thanks for this feature can you check our ai hints this could help to get the best out of it

No problem @ReneWerner87 !

In regards to the Code Rabbit hints, I've added a modified simple SSE example using the example the AI hints gave to the `whats_new.md` file.

The hints also mention about checking the return value of `fmt.Fprintf(w, ...)`. It appears that `fmt.Fprintf` (under normal circumstances if I'm not mistaken) should not return any errors, including if the connection is disconnected. The method `Flush()` does return an error in this on a disconnection. Should we still include error checking for `fmt.Fprintf` calls in the examples?

Thanks again for the suggestions and comments!

**Review by efectn — APPROVED** [review]

File: ctx.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaby) [reply ctx.go:1676]: Yes

**Comment by gaby** [comment]
@grivera64 I think you have to run `make generate` since the ctx.go changed

**Comment by grivera64** [comment]
> @grivera64 I think you have to run `make generate` since the ctx.go changed

Running `make generate` doesn't provide any new diffs to commit:

```sh
fiber (feature/add-buffered-streaming-support) $ make generate
go install github.com/tinylib/msgp@latest
go install github.com/vburenin/ifacemaker@975a95966976eeb2d4365a7fb236e274c54da64c
go generate ./...
fiber (feature/add-buffered-streaming-support) $ git status
On branch feature/add-buffered-streaming-support
Your branch is up to date with 'origin/feature/add-buffered-streaming-support'.

nothing to commit, working tree clean
```

Is this fine?

**Comment by grivera64** [comment]
> 👍 Just one comment. LGTM

@gaby For sure thanks for the review!

With regards to the recipes/example comment, should I make an issue on [gofiber/recipes](https://github.com/gofiber/recipes) to update the original [SSE example](https://github.com/gofiber/recipes/tree/master/sse) to use `c.SendStreamWriter()`?

**Comment by efectn** [comment]
> > 👍 Just one comment. LGTM
> 
> @gaby For sure thanks for the review!
> 
> With regards to the recipes/example comment, should I make an issue on [gofiber/recipes](https://github.com/gofiber/recipes) to update the original [SSE example](https://github.com/gofiber/recipes/tree/master/sse) to use `c.SendStreamWriter()`?

Yes, you can

**Comment by gaby** [comment]
> > @grivera64 I think you have to run `make generate` since the ctx.go changed
> 
> Running `make generate` doesn't provide any new diffs to commit:
> 
> ```shell
> fiber (feature/add-buffered-streaming-support) $ make generate
> go install github.com/tinylib/msgp@latest
> go install github.com/vburenin/ifacemaker@975a95966976eeb2d4365a7fb236e274c54da64c
> go generate ./...
> fiber (feature/add-buffered-streaming-support) $ git status
> On branch feature/add-buffered-streaming-support
> Your branch is up to date with 'origin/feature/add-buffered-streaming-support'.
> 
> nothing to commit, working tree clean
> ```
> 
> Is this fine?

Yes, that's fine 💪

**Pull request merged**